common/src/api-types.ts

```diff
@@ -0,0 +1,122 @@
+import * as t from 'io-ts'
+
+type Field<
+  Base,
+  T extends Base | undefined,
+  K extends keyof Base,
+  Fallback = undefined
+> = T extends {
+  [k in K]: Base[K]
+}
+  ? T[K]
+  : Fallback
+type TypeOf<T> = T extends t.Any ? t.TypeOf<T> : {}
+
+export namespace Schema {
+  export interface Request {
+    readonly parameters?: t.Any
+    readonly query?: t.Any
+    readonly body?: t.Any
+  }
+
+  export interface Response {
+    readonly body?: t.Any
+  }
+
+  export interface Method {
+    readonly request?: Request
+    readonly response?: Response
+  }
+
+  export type MethodName = 'all' | 'get' | 'post' | 'put' | 'delete' | 'patch' | 'options' | 'head'
+  export type Methods = Readonly<Partial<Record<MethodName, Method>>>
+
+  export type Path = `/${string}`
+  export type RootPath = '/'
+  export type Children = Readonly<Partial<Record<Exclude<Path, RootPath>, Endpoint>>>
```

Comment:
I'm not sure actually, it's to signify that not every `Path` is guaranteed to be in `.children`. I'm not sure if this happens by default in `Record`, I didn't check

Reply:
Did some looking into it; the code works without the `Partial` here. However, if you remove `Partial` from the definition of the similar `Methods` type, it won't work. If your type has a finite number of values (like `MethodName`), then Typescript expects every value to be a property in the corresponding record. Removing the `Partial` in the definition of `Children` works because there are infinitely many possible `Path`s.

tl;dr: if you want a record type where all the keys are optional, `Partial<Record<K, V>>` _always_ works. Depending on `K`, you can sometimes get away with just doing `Record<K, V>`. To be safe, stick to always wrapping a `Partial` around your type, so that way your type definition works regardless of `K`.